ballerina/bytes_to_sse_stream_generator.bal

```diff
@@ -0,0 +1,168 @@
+    DATA = "data"
+};
+
+class BytesToEventStreamGenerator {
+    private final stream<byte[], io:Error?> byteStream;
+    private boolean isClosed = false;
+    private byte[] lookaheadBuffer = [];
+
+    isolated function init(stream<byte[], io:Error?>? byteStream) {
+        self.byteStream = byteStream is () ? (<byte[][]>[]).toStream() : byteStream;
+    }
+
+    public isolated function next() returns record {|SseEvent value;|}|error? {
+        do {
+            string? sseEvent = check self.readUntilDoubleLineBreaks();
+            if sseEvent is () {
+                check self.close();
+                return ();
+            }
+            return {value: check parseSseEvent(sseEvent)};
+        } on fail error e {
+            log:printError("Failed to construct SseEvent", e);
+            return e;
+        }
+    }
+
+    public isolated function close() returns error? {
+        check self.byteStream.close();
+        self.isClosed = true;
+    }
+
+    private isolated function readUntilDoubleLineBreaks() returns string|error? {
+        byte[] buffer = [];
+        byte prevByte = 0;
+        byte? currentByte = ();
+        boolean foundCariageReturnWithNewLine = false;
+        while !self.isClosed {
+            currentByte = check self.getNextByte();
+            if currentByte is () {
+                return ();
+            }
+            if foundCariageReturnWithNewLine && currentByte == CARRIAGE_RETURN {
+                // Lookahead for newline
+                byte? nextByte = check self.getNextByte();
+                if nextByte is () {
+                    return ();
+                }
+                if nextByte == LINE_FEED {
+                    buffer.push(currentByte);
+                    buffer.push(nextByte);
+                    return string:fromBytes(buffer);
+                }
+                // Store char in lookahead buffer if not newline
+                self.lookaheadBuffer.push(nextByte);
+            }
+            foundCariageReturnWithNewLine = false;
+            if (currentByte == LINE_FEED || currentByte == CARRIAGE_RETURN) && prevByte == currentByte {
+                buffer.push(currentByte);
+                return string:fromBytes(buffer);
+            }
+            if currentByte == LINE_FEED && prevByte == CARRIAGE_RETURN {
+                foundCariageReturnWithNewLine = true;
+            }
+            buffer.push(currentByte);
+            prevByte = currentByte;
+        }
+        return ();
+    }
+
+    # Reads next byte from the lookahead buffer if data is available, otherwise read from the byte stream
+    # + return - A `byte?` on success, `error` on failure
+    private isolated function getNextByte() returns byte|error? {
+        if self.lookaheadBuffer.length() > 0 {
+            return self.lookaheadBuffer.shift();
+        }
+        record {byte[] value;}? nextValue = check self.byteStream.next();
+        return nextValue is () ? () : nextValue.value[0];
```

Comment:
can't `nextValue.value` be an empty byte array?

Reply:
No, it can't be.